### PR TITLE
[tune] Refactor syncer

### PR DESCRIPTION
**Pull request by ujvl**
Refactors ugly code in log_sync and syncer.

This PR is a dependency to #6376.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: python/ray/tune/sync_client.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+NOOP = FunctionBasedClient(noop, noop)
+
+
+class CommandBasedClient(SyncClient):
+    def __init__(self,
+                 sync_up_template,
+                 sync_down_template,
+                 delete_template=noop_template):
+        """Syncs between two directories with the given command.
```
Comment (richardliaw) [inline python/ray/tune/sync_client.py]: newline after first line of docstring

File: python/ray/tune/sync_client.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        self.logfile = tempfile.NamedTemporaryFile(
+            prefix="log_sync", dir=logdir, suffix=".log", delete=False)
+
+    def sync_up(self, source, target):
+        return self._execute(self.sync_up_template, source, target)
+
+    def sync_down(self, source, target):
+        return self._execute(self.sync_down_template, source, target)
+
+    def delete(self, target):
```
Comment (richardliaw) [inline python/ray/tune/sync_client.py]: Why does it make sense that the SyncClient also handles deletes?

File: python/ray/tune/syncer.py [hunk 3]
```diff
@@ -303,32 +261,14 @@ def get_log_syncer(local_dir, remote_dir=None, sync_function=None):
     elif not remote_dir:
         sync_client = NOOP
     elif sync_function:
-        sync_client = _get_sync_client(sync_function)
+        sync_client = get_sync_client(sync_function)
     else:
-        sync_up = log_sync_template()
-        sync_down = log_sync_template(options="--remove-source-files")
```
Comment (richardliaw) [inline python/ray/tune/syncer.py]: is the remove-source-files no longer used?

File: python/ray/tune/sync_client.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+                " to be installed".format(GS_PREFIX))
+        template = "gsutil rsync -r {source} {target}"
+        delete_template = "gsutil rm -r {target}"
+    else:
+        raise ValueError("Upload uri must start with one of: {}"
+                         "".format(ALLOWED_REMOTE_PREFIXES))
+    return CommandBasedClient(template, template, delete_template)
+
+
+class SyncClient(object):
```
Comment (richardliaw) [inline python/ray/tune/sync_client.py]: Can you doc this?

File: python/ray/tune/sync_client.py (hunk 2 above)
Comment (ujvl) [reply python/ray/tune/sync_client.py]: Hmm I don't have a strong reason for why/why not, just seems intuitive to me and is the simplest way to implement. lmk if there's another place to put this functionality that makes more sense.

File: python/ray/tune/syncer.py (hunk 3 above)
Comment (ujvl) [reply python/ray/tune/syncer.py]: oh ill revert this here it should go in the next PR, but the answer to your question is yes—I'll add an explanation for this in the other PR as well.

File: python/ray/tune/sync_client.py (hunk 2 above)
Comment (richardliaw) [reply python/ray/tune/sync_client.py]: It seems to overload this class with behavior that is not related to syncing. How is the delete function even used? Is this behavior actually used in Tune yet?  If not, we should take it out. Can approve and merge  once so.

Comment (ujvl) [reply python/ray/tune/sync_client.py]: done. i'll add it to the other PR and we can discuss it there

File: python/ray/tune/syncer.py [hunk 5]
```diff
@@ -257,37 +234,18 @@ def get_cloud_syncer(local_dir, remote_dir=None, sync_function=None):
...
-                        "".format(ALLOWED_REMOTE_PREFIXES))
-
+    sync_client = get_cloud_sync_client(remote_dir)
+    _syncers[key] = Syncer(local_dir, remote_dir, sync_client)
     return _syncers[key]
 
 
-def get_log_syncer(local_dir, remote_dir=None, sync_function=None):
-    """Returns a log Syncer.
+def get_syncer(local_dir, remote_dir=None, sync_function=None):
```
Comment (richardliaw) [inline python/ray/tune/syncer.py]: nit: get_node_syncer

**Review by richardliaw — APPROVED** [review]
one renaming nit

**Comment by richardliaw** [comment]
jenkins test tune

**Comment by ujvl** [comment]
jenkins test tune

**Pull request merged**